x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/Warnings.java

```diff
@@ -10,33 +10,37 @@
 import org.elasticsearch.xpack.ql.tree.Source;
 
 import static org.elasticsearch.common.logging.HeaderWarning.addWarning;
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
 
 /**
  * Utilities to collect warnings for running an executor.
  */
 public class Warnings {
     static final int MAX_ADDED_WARNINGS = 20;
 
-    private final Source source;
+    private final String location;
+    private final String first;
 
     private int addedWarnings;
 
     public Warnings(Source source) {
-        this.source = source;
+        location = format("Line {}:{}: ", source.source().getLineNumber(), source.source().getColumnNumber());
+        first = format(
+            null,
+            "{}evaluation of [{}] failed, treating result as null. Only first {} failures recorded.",
+            location,
+            source.text(),
+            MAX_ADDED_WARNINGS
```

Comment:
Nit: Hm, the impact of that may be negligible, but this refactoring means that we eagerly create the formatted `first` warning message and the `location` string whenever a `Warnings` instance is created. May be slightly better to keep the logic lazy as it was.

Reply:
> this refactoring means that we eagerly create the formatted first warning message and the location string whenever a Warnings instance is created.

Indeed.

> May be slightly better to keep the logic lazy as it was.

Why? Memory-consumption concerns?
The change adds location information to the exception (registering) line as well. So the `location` needs to be built at construction time anyways (or at least the line/column recorded). The `first` could be deferred, but otoh, that'd be in the data handling path (not that it would matter much, though).

Reply:
Preface: please treat this comment as optional, it's not a big deal either way :)

As to why: Less about memory consumption, more that we're adding multiple calls to `format` now whenever an evaluator is created, as evaluators come with a `Warnings` instance; for an expression/function-heavy query, this might be (very, very, very slightly) affecting performance. For the happy path where no warning is issued, neither the `first` string nor the `location` string need to be built from the source; in the unhappy path, of course, the location is needed for every warning and then caching the location string is faster, but path should be hit less often.

Again, probably negligible so simplicity is more important than performance concerns. Just pointed it out because the previous approach (which only kept track of the source) looked simple enough.